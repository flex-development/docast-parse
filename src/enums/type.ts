/**
 * @file Enums - Type
 * @module docast/enums/Type
 */

/**
 * Node types.
 *
 * @see https://github.com/syntax-tree/unist#type
 *
 * @enum {Lowercase<string>}
 */
enum Type {
  BLOCK_TAG = 'block-tag',
  COMMENT = 'comment',
  IMPLICIT_DESCRIPTION = 'implicit-description',
  INLINE_TAG = 'inline-tag',
  ROOT = 'root'
}

export default Type
