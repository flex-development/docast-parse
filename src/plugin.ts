/**
 * @file Plugin
 * @module docast-parse/plugin
 */

import type { Root } from '@flex-development/docast'
import { fromDocs, type Options } from '@flex-development/docast-util-from-docs'
import type { Nilable, Optional } from '@flex-development/tutils'
import type { Extension as MdastExtension } from 'mdast-util-from-markdown'
import type { Extension as MicromarkExtension } from 'micromark-util-types'
import type { Data, Plugin, Processor } from 'unified'
import type { VFile } from 'vfile'

declare module 'unified' {
  interface Data {
    /**
     * Markdown extensions to change how micromark tokens are converted to
     * nodes.
     *
     * @see {@linkcode MdastExtension}
     * @see https://github.com/syntax-tree/mdast-util-from-markdown#list-of-extensions
     */
    fromMarkdownExtensions?: Optional<MdastExtension[]>

    /**
     * Micromark extensions to change how markdown is parsed.
     *
     * @see {@linkcode MicromarkExtension}
     * @see https://github.com/micromark/micromark#extensions
     */
    micromarkExtensions?: Optional<MicromarkExtension[]>
  }

  interface Settings extends Options {}
}

/**
 * Configure the docblock parser.
 *
 * @see {@linkcode Options}
 * @see {@linkcode Processor}
 *
 * @this {Processor}
 *
 * @param {Nilable<Options>?} [options] - Configuration options
 * @return {void} Nothing
 */
function plugin(this: Processor, options?: Nilable<Options>): void {
  /**
   * Processor data.
   *
   * @const {Data} data
   */
  const data: Data = this.data()

  // initialize extensions
  data.fromMarkdownExtensions ??= []
  data.micromarkExtensions ??= []

  // initialize options
  options ??= {}
  options.mdastExtensions ??= []
  options.micromarkExtensions ??= []

  // configure extensions
  data.fromMarkdownExtensions.push(...options.mdastExtensions)
  data.micromarkExtensions.push(...options.micromarkExtensions)

  /**
   * Docblock parser.
   *
   * @param {string} document - Source document
   * @param {VFile} file - Source file
   * @return {Root} docast tree
   */
  const parser = (document: string, file: VFile): Root => {
    return fromDocs(String(file), {
      ...this.data('settings'),
      ...options,
      mdastExtensions: data.fromMarkdownExtensions,
      micromarkExtensions: data.micromarkExtensions
    })
  }

  return void (this.parser = parser)
}

/**
 * Add support for docblock parsing.
 *
 * @see {@linkcode Options}
 * @see {@linkcode Plugin}
 * @see {@linkcode Root}
 *
 * @this {Processor}
 *
 * @param {Nilable<Options>?} [options] - Configuration options
 * @return {void} Nothing
 */
const docastParse: Plugin<[Nilable<Options>?], string, Root> = plugin

export default docastParse
