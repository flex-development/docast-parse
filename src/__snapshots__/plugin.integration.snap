// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:plugin > empty document > should configure parser 1`] = `root[0]`;

exports[`integration:plugin > non-empty document > should configure parser 1`] = `
root[9]
├─0 comment[3] (1:1-5:4, 0-142)
│   │ code: null
│   ├─0 blockTag<file>[1] (2:4-2:38, 7-41)
│   │   │ tag: "@file"
│   │   └─0 text "Fixtures - FibonacciSequence" (2:10-2:38, 13-41)
│   ├─1 blockTag<module>[1] (3:4-3:38, 45-79)
│   │   │ tag: "@module"
│   │   └─0 text "fixtures/FibonacciSequence" (3:12-3:38, 53-79)
│   └─2 blockTag<see>[1] (4:4-4:59, 83-138)
│       │ tag: "@see"
│       └─0 text "https://codewars.com/kata/55695bc4f75bbaea5100016b" (4:9-4:59, 88-138)
├─1 comment[2] (7:1-19:4, 144-434)
│   │ code: null
│   ├─0 description[4] (8:4-16:7, 151-385)
│   │   ├─0 paragraph[1] (8:4-8:32, 151-179)
│   │   │   └─0 text "Fibonacci sequence iterator." (8:4-8:32, 151-179)
│   │   ├─1 break (8:32-9:1, 179-180)
│   │   ├─2 break (9:3-10:1, 182-183)
│   │   │     data: {"blank":true}
│   │   └─3 containerDirective<info>[2] (10:4-16:7, 186-385)
│   │       │ attributes: {}
│   │       ├─0 paragraph[3] (11:4-12:37, 197-308)
│   │       │   ├─0 text "A fibonacci sequence starts with two " (11:4-11:41, 197-234)
│   │       │   ├─1 inlineCode "1" (11:41-11:44, 234-237)
│   │       │   └─2 text "s. Every element afterwards is the sum of the two previous elements:" (11:44-12:37, 237-308)
│   │       └─1 code "1, 1, 2, 3, 5, 8, 13, ..., 89, 144, 233, 377, ..." (13:4-15:7, 312-378)
│   │             lang: "txt"
│   │             meta: null
│   └─1 blockTag<implements>[1] (18:4-18:42, 392-430)
│       │ tag: "@implements"
│       └─0 typeExpression "Iterator<number, number>" (18:16-18:42, 404-430)
├─2 comment[4] (21:3-27:6, 499-603)
│   │ code: null
│   ├─0 description[1] (22:6-22:35, 508-537)
│   │   └─0 paragraph[1] (22:6-22:35, 508-537)
│   │       └─0 text "First managed sequence value." (22:6-22:35, 508-537)
│   ├─1 blockTag<public>[0] (24:6-24:13, 548-555)
│   │     tag: "@public"
│   ├─2 blockTag<instance>[0] (25:6-25:15, 561-570)
│   │     tag: "@instance"
│   └─3 blockTag<member>[2] (26:6-26:27, 576-597)
│       │ tag: "@member"
│       ├─0 typeExpression "number" (26:14-26:22, 584-592)
│       └─1 text "fib1" (26:23-26:27, 593-597)
├─3 comment[4] (30:3-36:6, 629-734)
│   │ code: null
│   ├─0 description[1] (31:6-31:36, 638-668)
│   │   └─0 paragraph[1] (31:6-31:36, 638-668)
│   │       └─0 text "Second managed sequence value." (31:6-31:36, 638-668)
│   ├─1 blockTag<public>[0] (33:6-33:13, 679-686)
│   │     tag: "@public"
│   ├─2 blockTag<instance>[0] (34:6-34:15, 692-701)
│   │     tag: "@instance"
│   └─3 blockTag<member>[2] (35:6-35:27, 707-728)
│       │ tag: "@member"
│       ├─0 typeExpression "number" (35:14-35:22, 715-723)
│       └─1 text "fib2" (35:23-35:27, 724-728)
├─4 comment[4] (39:3-45:6, 760-854)
│   │ code: null
│   ├─0 description[1] (40:6-40:25, 769-788)
│   │   └─0 paragraph[1] (40:6-40:25, 769-788)
│   │       └─0 text "Max sequence value." (40:6-40:25, 769-788)
│   ├─1 blockTag<private>[0] (42:6-42:14, 799-807)
│   │     tag: "@private"
│   ├─2 blockTag<instance>[0] (43:6-43:15, 813-822)
│   │     tag: "@instance"
│   └─3 blockTag<member>[2] (44:6-44:26, 828-848)
│       │ tag: "@member"
│       ├─0 typeExpression "number" (44:14-44:22, 836-844)
│       └─1 text "max" (44:23-44:26, 845-848)
├─5 comment[2] (48:3-52:6, 882-1015)
│   │ code: null
│   ├─0 description[1] (49:6-49:47, 891-932)
│   │   └─0 paragraph[1] (49:6-49:47, 891-932)
│   │       └─0 text "Create a new fibonacci sequence iterator." (49:6-49:47, 891-932)
│   └─1 blockTag<param>[2] (51:6-51:72, 943-1009)
│       │ tag: "@param"
│       ├─0 typeExpression "number" (51:13-51:21, 950-958)
│       └─1 text "[max=Number.MAX_SAFE_INTEGER] - Max sequence value" (51:22-51:72, 959-1009)
├─6 comment[4] (58:3-65:6, 1142-1279)
│   │ code: null
│   ├─0 description[1] (59:6-59:24, 1151-1169)
│   │   └─0 paragraph[1] (59:6-59:24, 1151-1169)
│   │       └─0 text "Iterable protocol." (59:6-59:24, 1151-1169)
│   ├─1 blockTag<public>[0] (61:6-61:13, 1180-1187)
│   │     tag: "@public"
│   ├─2 blockTag<instance>[0] (62:6-62:15, 1193-1202)
│   │     tag: "@instance"
│   └─3 blockTag<return>[2] (64:6-64:66, 1213-1273)
│       │ tag: "@return"
│       ├─0 typeExpression "IterableIterator<number>" (64:14-64:40, 1221-1247)
│       └─1 text "Current sequence iterator" (64:41-64:66, 1248-1273)
├─7 comment[4] (70:3-77:6, 1360-1524)
│   │ code: null
│   ├─0 description[1] (71:6-71:51, 1369-1414)
│   │   └─0 paragraph[1] (71:6-71:51, 1369-1414)
│   │       └─0 text "Get the next value in the fibonacci sequence." (71:6-71:51, 1369-1414)
│   ├─1 blockTag<public>[0] (73:6-73:13, 1425-1432)
│   │     tag: "@public"
│   ├─2 blockTag<instance>[0] (74:6-74:15, 1438-1447)
│   │     tag: "@instance"
│   └─3 blockTag<return>[2] (76:6-76:66, 1458-1518)
│       │ tag: "@return"
│       ├─0 typeExpression "IteratorResult<number, number>" (76:14-76:46, 1466-1498)
│       └─1 text "Next sequence value" (76:47-76:66, 1499-1518)
└─8 comment[2] (79:5-83:8, 1579-1659)
    │ code: null
    ├─0 description[1] (80:8-80:33, 1590-1615)
    │   └─0 paragraph[1] (80:8-80:33, 1590-1615)
    │       └─0 text "Temporary sequence value." (80:8-80:33, 1590-1615)
    └─1 blockTag<const>[2] (82:8-82:29, 1630-1651)
        │ tag: "@const"
        ├─0 typeExpression "number" (82:15-82:23, 1637-1645)
        └─1 text "value" (82:24-82:29, 1646-1651)
`;
